x/upgrade/client/cli/tx.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/cosmos/cosmos-sdk/x/gov"
+	upgrade "github.com/cosmos/cosmos-sdk/x/upgrade/internal/types"
+)
+
+const (
+	// TimeFormat specifies ISO UTC format for submitting the upgrade-time for a new upgrade proposal
+	TimeFormat = "2006-01-02T15:04:05Z"
+
+	FlagUpgradeName   = "upgrade-name"
+	FlagUpgradeHeight = "upgrade-height"
+	FlagUpgradeTime   = "upgrade-time"
+	FlagUpgradeInfo   = "upgrade-info"
+)
+
+func parseSubmitArgs(cmd *cobra.Command, cdc *codec.Codec) (gov.Content, error) {
+	title, err := cmd.Flags().GetString(cli.FlagTitle)
+	if err != nil {
+		return nil, err
+	}
+
+	description, err := cmd.Flags().GetString(cli.FlagDescription)
+	if err != nil {
+		return nil, err
+	}
+
+	name, err := cmd.Flags().GetString(FlagUpgradeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(name) == 0 {
+		name = title
+	}
+
+	height, err := cmd.Flags().GetInt64(FlagUpgradeHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	timeStr, err := cmd.Flags().GetString(FlagUpgradeTime)
+	if err != nil {
+		return nil, err
+	}
+
+	if height != 0 {
+		if len(timeStr) != 0 {
+			return nil, fmt.Errorf("only one of --upgrade-time or --upgrade-height should be specified")
+		}
+	}
+
+	var upgradeTime time.Time
+	if len(timeStr) != 0 {
+		upgradeTime, err = time.Parse(TimeFormat, timeStr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	info, err := cmd.Flags().GetString(FlagUpgradeInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	content := upgrade.NewSoftwareUpgradeProposal(title, description,
+		upgrade.Plan{Name: name, Time: upgradeTime, Height: height, Info: info})
+	return content, nil
+}
+
+// GetCmdSubmitUpgradeProposal implements a command handler for submitting a software upgrade proposal transaction.
+func GetCmdSubmitUpgradeProposal(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "software-upgrade --upgrade-name [name] (--upgrade-height [height] | --upgrade-time [time]) (--upgrade-info [info]) [flags]",
+		Args:  cobra.ExactArgs(0),
+		Short: "Submit a software upgrade proposal",
+		Long: "Submit a software upgrade along with an initial deposit.\n" +
+			"Please specify a unique name and height OR time for the upgrade to take effect.\n" +
+			"You may include info to reference a binary download link, in a format compatible with: https://github.com/regen-network/cosmosd",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			content, err := parseSubmitArgs(cmd, cdc)
+			if err != nil {
+				return err
+			}
+
+			txBldr := auth.NewTxBuilderFromCLI().WithTxEncoder(utils.GetTxEncoder(cdc))
+			cliCtx := context.NewCLIContext().WithCodec(cdc)
+			from := cliCtx.GetFromAddress()
+
+			depositStr, err := cmd.Flags().GetString(cli.FlagDeposit)
+			if err != nil {
+				return err
+			}
+			deposit, err := sdk.ParseCoins(depositStr)
+			if err != nil {
+				return err
+			}
+
+			msg := gov.NewMsgSubmitProposal(content, deposit, from)
+			if err := msg.ValidateBasic(); err != nil {
```

Comment:
this is already checked by the AnteHandler. Any reason to have it here as a dup?

Reply:
This is checking it client side to provide an immediate error message, before sending it to the chain. It is a nicer user experience (and cleaner message)